Supporting mapml-proxy cascading

dromagnoli opened the pull request:
<!--Include a few sentences describing the overall goals for this Pull Request-->
  
<!-- Please help our volunteers reviewing this PR by completing the following items. 
Ask in a comment if you have troubles with any of them. -->

# Checklist

- [ ] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [ ] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [ ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [ ] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

<!--Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.-->

prushforth commented:
I have successfully used this branch with the CFS CWFIS WMS (GeoServer!), and I found it works well in Remote request mode, except the remote GetFeatureInfo request balks at including of a blank `cql_filter` paramter:

e.g. the following template 
`https://cwfis.cfs.nrcan.gc.ca/geoserver/wms?request=GetFeatureInfo&query_layers=fdr_current&crs=EPSG:3978&bbox={xmin},{ymin},{xmax},{ymax}&language=en&cql_filter=&version=1.3.0&transparent=true&service=WMS&layers=fdr_current&width={w}&x={i}&feature_count=50&y={j}&styles=&info_format=text/html&height={h}`

when filled out gives rise to an example query such as:

https://cwfis.cfs.nrcan.gc.ca/geoserver/wms?request=GetFeatureInfo&query_layers=fdr_current&crs=EPSG:3978&bbox=-1150222.572178498,-634225.7217853889,863260.6214545518,1688820.5909739062&language=en&cql_filter=&version=1.3.0&transparent=true&service=WMS&layers=fdr_current&width=761&x=381&feature_count=50&y=439&styles=&info_format=text/html&height=878

I find that if I edit the above URL and remove the cql_filter parameter entirely, it works ok:

https://cwfis.cfs.nrcan.gc.ca/geoserver/wms?request=GetFeatureInfo&query_layers=fdr_current&crs=EPSG:3978&bbox=-1150222.572178498,-634225.7217853889,863260.6214545518,1688820.5909739062&language=en&version=1.3.0&transparent=true&service=WMS&layers=fdr_current&width=761&x=381&feature_count=50&y=439&styles=&info_format=text/html&height=878

Maybe this is actually a bug in GeoServer cql_filter handling?  Anyway, I will see if there are other cases I come across.

dromagnoli commented:
PR updated with WMTS tests and getFeatureInfo with empty cql_filter being properly handled.

prushforth commented:
Just returning to playing with this, and maybe it's something odd about the remote WMS, but I can't get it to proxy correctly, it always seems to truncate the "_en" from the remote URL.  The remote WMS capabilities is [here](https://maps.geogratis.gc.ca/wms/toporama_en?request=GetCapabilities&service=WMS&version=1.1.1)

The remote layer name to proxy is "WMS-Toporama", which is like a layer group afaict on the remote wms.  
![image](https://github.com/user-attachments/assets/8fc2da05-3afe-4763-9b90-9e6f4d7ec10f)



The client proxy always comes back with a 404, because it uses an incorrect URL in the tref attribute of the \<map-link>: 
![image](https://github.com/user-attachments/assets/1f601e4e-1bec-47b8-a430-e0ce93153105)

prushforth commented on src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLRequestMangler.java:
```diff
@@ -0,0 +1,607 @@
+            }
+            // Optionally, add the context path if needed
+            String path = url.getPath();
+            int contextPathEnd = path.indexOf("/", 1);
```
This is the culprit, I think

prushforth replied:
Yes, thanks that seems to work now.

dromagnoli commented:
> Ideally, creation (at all) of the <map-link rel="query"> would be conditional on if the remote layer is queryable.
> 
> IIRC GeoServer itself can cascade remote queries if the remote WMS supports some flavour of GML. Maybe we could change that when the Remote proxy option for MapML is selected, to `text/mapml`, `text/html` or `text/plain` availability? If none is available we would not create a <map-link rel="query" at all, as though the remote layer was not queryable in the first place.


I have updated the code so that it won't generate the query link if the remote server has none of these format supported.
I have also added a test.

prushforth commented:
Works well for MapML:CBMTILE with this WMS however. Probably due to the quality of the remove WMS, https://cwfis.cfs.nrcan.gc.ca/geoserver/public/wms?service=WMS&version=1.3.0&request=GetCapabilities

dromagnoli commented:
Fixed the NPE on missing GetFeatureInfo from remote capabilities document

dromagnoli commented:
Closing this in favor of the official PR:
https://github.com/geoserver/geoserver/pull/7930
